vowpalwabbit/slim/include/vw_slim_predict.h

```diff
@@ -0,0 +1,451 @@
+        }
+      }
+      else if (find_opt_float(_command_line_arguments, "--epsilon", _epsilon))
+        _exploration = vw_predict_exploration::epsilon_greedy;
+      else
+        return E_VW_PREDICT_ERR_CB_EXPLORATION_MISSING;
+    }
+
+    // VW style check_sum validation
+    uint32_t check_sum_computed = mp.checksum();
+
+    // perform check sum check
+    uint32_t check_sum_len;
+    RETURN_ON_FAIL((mp.read<uint32_t, false>("check_sum_len", check_sum_len)));
+    if (check_sum_len != sizeof(uint32_t))
+      return E_VW_PREDICT_ERR_INVALID_MODEL;
+
+    uint32_t check_sum;
+    RETURN_ON_FAIL((mp.read<uint32_t, false>("check_sum", check_sum)));
+
+    if (check_sum_computed != check_sum)
+      return E_VW_PREDICT_ERR_INVALID_MODEL_CHECK_SUM;
+
+    if (_command_line_arguments.find("--cb_adf") != std::string::npos)
+    {
+      RETURN_ON_FAIL(mp.skip(sizeof(uint64_t)));  // cb_adf.cc: event_sum
+      RETURN_ON_FAIL(mp.skip(sizeof(uint64_t)));  // cb_adf.cc: action_sum
+    }
+
+    // gd.cc: save_load
+    bool gd_resume;
+    RETURN_ON_FAIL(mp.read("resume", gd_resume));
+    if (gd_resume)
+      return E_VW_PREDICT_ERR_GD_RESUME_NOT_SUPPORTED;
+
+    // read sparse weights into dense
+    uint64_t weight_length = (uint64_t)1 << _num_bits;
+    _stride_shift = (uint32_t)ceil_log_2(num_weights);
+
+    RETURN_ON_FAIL(mp.read_weights<W>(_weights, _num_bits, _stride_shift));
+
+    // TODO: check that permutations is not enabled (or parse it)
+
+    _model_loaded = true;
+
+    return S_VW_PREDICT_OK;
+  }
+
+  /**
+   * @brief True if the model describes a contextual bandit (cb) model using action dependent features (afd)
+   *
+   * @return true True if contextual bandit predict method can be used.
+   * @return false False if contextual bandit predict method cannot be used.
+   */
+  bool is_cb_explore_adf() { return _command_line_arguments.find("--cb_explore_adf") != std::string::npos; }
+
+  /**
+   * @brief True if the model describes a cost sensitive one-against-all (csoaa). This is also true for cb_explore_adf
+   * models, as they are reduced to csoaa.
+   *
+   * @return true True if csoaa predict method can be used.
+   * @return false False if csoaa predict method cannot be used.
+   */
+  bool is_csoaa_ldf() { return _command_line_arguments.find("--csoaa_ldf") != std::string::npos; }
+
+  /**
+   * @brief Predicts a score (as in regression) for the provided example.
+   *
+   * Regular regression with support for constant feature (bias term) and interactions
+   *
+   * @param ex The example to get the prediction for.
+   * @param score The output score produced by the model.
+   * @return int Returns 0 (S_VW_PREDICT_OK) if succesful, otherwise one of the error codes (see E_VW_PREDICT_ERR_*).
+   */
+  int predict(example_predict& ex, float& score)
+  {
+    if (!_model_loaded)
+      return E_VW_PREDICT_ERR_NO_MODEL_LOADED;
+
+    std::unique_ptr<namespace_copy_guard> ns_copy_guard;
```

Comment:
Why is this a unique_ptr? Looks like it could just be stack allocated?

Reply:
Yes.  I remember trying to make this stack allocated.  I ran into some issues around references.  I don't recall what now.  

---
In reply to: [304976027](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304976027) [](ancestors = 304976027)